### PR TITLE
add support for PKCS#11 tokens

### PR DESCRIPTION
**Pull request by jluebbe**
This allows using HSMs or other tokens to store the signing keys.

### CODE REVIEW

File: src/signature.c [hunk 1]
```diff
@@ -20,7 +21,84 @@ void signature_init(void)
...
-static EVP_PKEY *load_key(const gchar *keyfile, GError **error)
+static ENGINE *get_pkcs11_engine(GError **error)
+{
+	static ENGINE *e = NULL;
+	unsigned long err;
+	const gchar *data;
+	const gchar *env;
+	int flags;
+
+	if (e)
```
Comment (ejoerns) [inline src/signature.c]: Should not happen as you just set it to NULL...

File: src/signature.c [hunk 2]
```diff
@@ -20,7 +21,84 @@ void signature_init(void)
 	ERR_load_crypto_strings();
 }
 
-static EVP_PKEY *load_key(const gchar *keyfile, GError **error)
+static ENGINE *get_pkcs11_engine(GError **error)
```
Comment (ejoerns) [inline src/signature.c]: Please add a `g_return_val_if_fail(..)` for `error`

File: src/signature.c [hunk 3]
```diff
@@ -57,14 +135,59 @@ static EVP_PKEY *load_key(const gchar *keyfile, GError **error)
...
+
+	g_return_val_if_fail(url != NULL, NULL);
+	g_return_val_if_fail(error == NULL || *error == NULL, NULL);
+
+	e = get_pkcs11_engine(&ierror);
+	if (e == NULL) {
+		g_propagate_error(error, ierror);
+		goto out;
+	}
+
```
Comment (ejoerns) [inline src/signature.c]: excess empty line

**Comment by ejoerns** [comment]
For now I've manually picked the lower 3 testing fixes which are anyway more ore less unrelated to the rest.
These fixes are really useful for testing and I came across the issues they fix several times only during the past few day.

Thus please rebase your changes onto master for the next round. Thanks!

File: docs/advanced.rst [hunk 4]
```diff
@@ -197,6 +197,99 @@ Example hook shell script code for above trust chain:
...
+For example, a bundle can be signed with a certificate and key available as
+``pkcs11:token=rauc;object=autobuilder-1``::
+
+  rauc bundle \
+    --cert='pkcs11:token=rauc;object=autobuilder-1' \
+    --key='pkcs11:token=rauc;object=autobuilder-1' \
+    <input-dir> <output-file>
+
+.. note::
+  Most PKCS#11 implementation require a PIN for signing operations.
```
[the remark this answers is not in the document]
Comment (jluebbe) [reply docs/advanced.rst]: Fixed.

File: src/signature.c (hunk 1 above)
Comment (jluebbe) [reply src/signature.c]: Fixed.

File: src/signature.c (hunk 2 above)
Comment (jluebbe) [reply src/signature.c]: Done.

File: src/signature.c (hunk 3 above)
Comment (jluebbe) [reply src/signature.c]: Removed.

**Pull request merged**